lib/Sema/CSGen.cpp

```diff
@@ -2262,47 +2305,93 @@ namespace {
         // ascribed type.
         Type subPatternType =
             getTypeForPattern(
-               cast<TypedPattern>(pattern)->getSubPattern(), locator);
+               cast<TypedPattern>(pattern)->getSubPattern(), locator,
+               Type(), bindPatternVarsOneWay);
         CS.addConstraint(
             ConstraintKind::Conversion, subPatternType, openedType, locator);
         return setType(openedType);
       }
 
       case PatternKind::Tuple: {
         auto tuplePat = cast<TuplePattern>(pattern);
+
+        // If there's an externally-imposed type, decompose it into element
+        // types so long as we have the right number of such types.
+        SmallVector<AnyFunctionType::Param, 4> externalEltTypes;
+        if (externalPatternType) {
+          AnyFunctionType::decomposeInput(externalPatternType,
+                                          externalEltTypes);
+
+          // If we have the wrong number of elements, we may not be able to
+          // provide more specific types.
+          if (tuplePat->getNumElements() != externalEltTypes.size()) {
+            externalEltTypes.clear();
+
+            // Implicit tupling.
+            if (tuplePat->getNumElements() == 1) {
+              externalEltTypes.push_back(
+                  AnyFunctionType::Param(externalPatternType));
+            }
+          }
+        }
+
         SmallVector<TupleTypeElt, 4> tupleTypeElts;
         tupleTypeElts.reserve(tuplePat->getNumElements());
         for (unsigned i = 0, e = tuplePat->getNumElements(); i != e; ++i) {
           auto &tupleElt = tuplePat->getElement(i);
-          Type eltTy = getTypeForPattern(tupleElt.getPattern(),
-                                         locator.withPathElement(
-                                           LocatorPathElt::TupleElement(i)));
+          Type externalEltType;
+          if (!externalEltTypes.empty())
+            externalEltType = externalEltTypes[i].getPlainType();
+
+          Type eltTy = getTypeForPattern(
+              tupleElt.getPattern(),
+              locator.withPathElement(LocatorPathElt::TupleElement(i)),
+              externalEltType,
+              bindPatternVarsOneWay);
           tupleTypeElts.push_back(TupleTypeElt(eltTy, tupleElt.getLabel()));
         }
+
         return setType(TupleType::get(tupleTypeElts, CS.getASTContext()));
       }
 
       case PatternKind::OptionalSome: {
+        // Remove an optional from the object type.
+        if (externalPatternType) {
+          if (Type objType = externalPatternType->getOptionalObjectType()) {
+            externalPatternType = objType;
+          } else {
+            Type objVar = CS.createTypeVariable(
+                CS.getConstraintLocator(locator), TVO_CanBindToNoEscape);
+            CS.addConstraint(
```

Comment:
Does it make sense to use `OptionalObject` constraint here instead? it seems like what this constraint is trying to do is to bind object type of the optional to `objVar`.

Reply:
Yes, that's what I'm trying to do. I had completely forgotten about `OptionalObject`, thanks!